Fixes outdated sensor data after reset

kellyguo11 opened the pull request:
# Description

This change adds a call to `update_articulations_kinematic()` after performing reset in an environment to ensure that non-render sensors are updated after performing reset.

## Type of change

<!-- As you go through the list, delete the ones that are not applicable. -->

- Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [ ] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

<!--
As you go through the checklist above, you can mark something as done by putting an x character in it

For example,
- [x] I have done this task
- [ ] I have not done this task
-->

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/manager_based_env.py:
```diff
@@ -251,6 +251,8 @@ def reset(self, seed: int | None = None, options: dict[str, Any] | None = None)
         indices = torch.arange(self.num_envs, dtype=torch.int64, device=self.device)
         self._reset_idx(indices)
 
+        # update articulation kinematics
```
I think this should be called after the event manager is called to reset transforms (so all the other managers also get this change)

> https://github.com/isaac-sim/IsaacLab/blob/5ab1f0bf57540fbfb6fddb562ca99612940c33a0/source/extensions/omni.isaac.lab/omni/isaac/lab/envs/manager_based_env.py#L363

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/manager_based_rl_env.py:
```diff
@@ -194,6 +194,8 @@ def step(self, action: torch.Tensor) -> VecEnvStepReturn:
         reset_env_ids = self.reset_buf.nonzero(as_tuple=False).squeeze(-1)
         if len(reset_env_ids) > 0:
             self._reset_idx(reset_env_ids)
+            # update articulation kinematics
```
Same comment as the manager based env.

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sim/simulation_context.py:
```diff
@@ -394,6 +394,14 @@ def get_setting(self, name: str) -> Any:
         """
         return self._settings.get(name)
 
+    def update_fabric_and_kinematics(self) -> None:
```
Hmm why not just put this inside the `render` call?

Mayankm96 replied:
Is it also needed for other cases?

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sim/simulation_context.py:
```diff
@@ -394,6 +394,14 @@ def get_setting(self, name: str) -> Any:
         """
         return self._settings.get(name)
 
+    def update_fabric_and_kinematics(self) -> None:
```
I am wondering if we should go for a more canonical terminology. Many simulators distinguish between kinematic and dynamics stepping through different terms. For example, Mujoco calls forward kinematics function as `forward` and the simulation step as `simulate`.
 
Something like this is more user-friendly.

```suggestion
    def forward(self) -> None:
```

Mayankm96 commented on source/extensions/omni.isaac.lab/docs/CHANGELOG.rst:
```diff
@@ -1,6 +1,15 @@
+Changed
+^^^^^^^
+
+* Added call to update articulation kinematics after reset to ensure updated states in non-rendering sensors.
```
I think it's a good practice to explain what was happening before this change was made. For instance, "Earlier, non-rendering sensors still used older transforms which corrupted the obtained readings."

This way it isn't just a change but also the motivation for the change that gets documented :)

Mayankm96 commented:
From an understanding standpoint, if a user sets robot state for articulation, isn't the base pose also updated in non-rendering sensors because the "views" capture the same memory. What does the update fabric do? 👀

Mayankm96 commented on source/extensions/omni.isaac.lab/omni/isaac/lab/sim/simulation_context.py:
```diff
@@ -464,11 +472,7 @@ def render(self, mode: RenderMode | None = None):
                 self.set_setting("/app/player/playSimulations", True)
         else:
             # manually flush the fabric data to update Hydra textures
-            if self._fabric_iface is not None:
```
Hm this would break non env workflows where users didn't need to call a forward function. Maybe we add a flag on whether forward was called and in-case it wasn't, we do it implicitly?

kellyguo11 replied:
This shouldn't actually change any logic right? It's just moving a piece of the code into a method.

kellyguo11 commented:
> From an understanding standpoint, if a user sets robot state for articulation, isn't the base pose also updated in non-rendering sensors because the "views" capture the same memory. What does the update fabric do? 👀
> 
> Also can we have some sort of test to make sure this behavior is always verified (for sensors)?

I think it's needed for capturing updates from the dof states and propagating those to the bodies. Added a test case for it.

nvcyc commented on source/extensions/omni.isaac.lab/omni/isaac/lab/envs/manager_based_env.py:
```diff
@@ -317,13 +319,13 @@ def reset_to(
         # set the state
         self.scene.reset_to(state, env_ids, is_relative=is_relative)
 
```
I think we also need to add `self.sim.forward()` here.
